### ISSUE TITLE
Proposal for modular assembly of services in container

### ISSUE THREAD

**Issue by mowens** [Swinject/Swinject#35]
# Proposal for modular assembly of services in container
Provide a simple mechanism to populate service definitions in a container that allows implementations to group related components together across frameworks/application

The current implementation has no opinion to how services are registered to the container which allows users to assembly their container(s) however you want. This is all well and good, but it would be nice to provide consumers of the framework an opinionated way to register their services. My proposal is to provide an `AssemblyType` and `Assembler` to provide such a feature. This approach would allow sane bootstrapping of the container (this concept is taking from is very vaguely taken from [Typhoon](https://github.com/appsquickly/Typhoon/wiki/Modularizing-Assemblies)

## Proposed Implementation
There are 2 components to the implementation: `AssemblyType` and `Assembler`. The `AssemblyType` provides an interface that can be implemented by classes/structs (depending on your use case) that is provided a container from elsewhere.

```
protocol AssemblyType {
    func assemble(container: Container)
}
```

The `Assembler` will take a list of `AssemblyType` instances and build the container. The `Assembler` will either:
a) Own the container
b) Create the container to be owned else where

The implementation would be really straight forward:

```
class Assembler {
   let container = Container()
   init(assemblies: [AssemblyType]) {
       for assembly in assemblies {
           assembly.assemble(container)
       }
   }
}
```

With this approach, one can separate their service definitions like so:

```
class ServiceAssembly: AssemblyType {
    func assemble(container: Container) {
        container.register(FooServiceType.self) { r in 
             FooService() 
        }
        container.register(BarServiceType.self) { r in 
            BarService() 
        }
    }
}

class ManagerAssembly: AssemblyType {
    func assemble(container: Container) {
        container.register(FooManagerType.self) { r in 
            FooManager(service: r.resolve(FooServiceType.self)!)
        }
        container.register(BarManagerType.self) { r in 
             BarManager(service: r.resolve(BarServiceType.self)!) 
        }
    }
}
```
Now like service definitions are grouped and we can build our container like so:
```
let assembler = Assembler(assemblies: [
    ServiceAssembly(),
    ManagerAssembly()
])
let container = assembler.container // will contain all registered services
```

Additionally, this provides an easy means to override service definitions with mocks using inheritance. Let's say we want to replace `FooService` with `MockFooService`:

```
class MockServiceAssembly: ServiceAssembly {
    override func assemble(container: Container) {
        super.assemble(container)
        
        container.register(FooServiceType.self) { r in 
             MockFooService() 
        }
    }
}
```

Then we would load our assembler with the `MockServiceAssembly` instead:
```
let assembler = Assembler(assemblies: [
    MockServiceAssembly(),
    ManagerAssembly()
])
```

**Comment by yoichitgy** [comment]
Thank you for your idea to add the feature! It is convenient and helpful for users who have used Typhoon or another DI framework supporting the assembly feature (if exists). The implementation plan is perfect👍

What I am thinking about new feature is structuring Swinject project. When I got PR #14, I thought extended features should be in another repository to organize the project. I was also thinking to make a repository for SwinjectStoryboard. However, Apple introduced their own [Swift Package Manager](https://github.com/apple/swift-package-manager) with [source layouts for modules](https://github.com/apple/swift-package-manager/blob/master/Documentation/SourceLayouts.md). Now I think it's good to follow their source layout convention like this:

    Swinject repository root
      |- Sources
      |   |- Core
      |   |- Assembly
      |   |- SwinjectStoryboard (if possible because it contains Objective-C code)
      |- Tests
          |- Core
          |- Assembly
          |- SwinjectStoryboard (if possible because it contains Objective-C code)

What do you think about the source layouts in a single repository or multiple repositories to add the assembly feature?

**Comment by mowens** [comment]
Good question. Be it that SPM is still so new (doesn't even support iOS, watchOS, or tvOS yet...) I believe we will see many improvements to its API over the next year. With that said the Source layout structure is probably a better idea compared to separate repos at this time. This would also make supporting CocoaPods very easy with subspecs (not so much for Carthage as now you have to maintain a separate build target for each sub-directory for each platform, but eh what can ya do)

Likely the submodules would need better naming to prevent conflicts as `Core` and `Assembly` are very generic. So would probably need to prefix Swinject to each submodule since you cannot specify this in the `Package` definition: `SwinjectCore`, `SwinjectAssembly` etc.. This is obviously very verbose but looks like a necessary requirement for SPM in its current state.

re: containing Objc code
This too I am very curious to see how this gets handled by Apple/Community. While Swift is awesome, many applications require Objc code, however I don't see SPM supporting Objc ever. Maybe there will be some plugin that will support xcodebuild to support Objc but who knows.

**Comment by yoichitgy** [comment]
Thanks Mike. now I'm confident to go with the single repo. I'm going to start renaming the source and test directories to try SPM. I won't make submodule directories for now to see more new information or discovery about SPM.

You can implement the assembly feature without modifying the existing code, can't you? Just to prepare the submodule layout in the future. I think, even if something should be added to `Container`, it can be implemented in a new file with an extension to `Container`.

I am curious to see the Objc things too. We should watch. Another curiosity is Apple has decided to implement Foundation in Swift. (Core)Foundation was implemented in C with toll free bridge to Objective-C. Probably UIKit will replace its dependency on CoreFoundation to Swift Foundation.

**Comment by mowens** [comment]
@yoichitgy Yea the modular concept I am proposing will not required any changes to the existing structure. :+1:

I will go ahead with the proposal and submit a feature branch. We can discuss merge approaches when the time comes :smiley:

**Comment by yoichitgy** [comment]
@mowens I added milestones to release the version 1.0.0 since we decided to manage Swinject in the single repo.
https://github.com/Swinject/Swinject/milestones

- **v1.0.0-beta.3:** Planned for Issue #38 (a small breaking change). If you have an opinion to #38, it's welcome.
- **v1.0.0:** Just wait for a week after the small breaking change to see no one complains.

The assembly feature can be included in `v1.0.0-beta.3` or `v1.1.0` when you finish the implementation😃
